sycl/doc/extensions/proposed/sycl_ext_oneapi_peer_access.asciidoc

```diff
@@ -0,0 +1,151 @@
+memory access between peer devices in a system.
+In particular, this allows one device to access USM Device allocations
+for a peer device. This extension does not apply to USM Shared allocations.
+Peer to peer capabilities are useful as they can provide
+access to a peer device's memory inside a compute kernel and optimized memory
+copies between peer devices.
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_PEER_ACCESS` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|Initial version of this extension.
+|===
+
+
+=== Peer to Peer (P2P) Memory Access APIs
+
+This extension adds support for mechanisms to query and enable support for
+direct memory access between peer devices in a system.
+In particular, this allows one device to directly access USM Device
+allocations for a peer device in the same context.
+Peer to peer capabilities are useful as they can provide access to a peer
+device's memory inside a compute kernel and also optimized memory copies between
+peer devices.
+
+This extension adds four new member functions to the device class, as described
+below.
+
+[source,c++]
+----
+namespace sycl   {
+namespace ext    {
+namespace oneapi {
+enum class peer_access {
+  access_supported,
+  access_enabled,
+  atomics_supported,
+};
+} // namespace oneapi
+} // namespace ext
+
+class device {
+public:
+  bool ext_oneapi_can_access_peer(const device &peer,
+                                  ext::oneapi::peer_access value =
+                                    ext::oneapi::peer_access::access_supported);
+  void ext_oneapi_enable_peer_access(const device &peer);
+  void ext_oneapi_disable_peer_access(const device &peer);
+};
+
+} // namespace sycl
+----
+
+The semantics of the new functions are:
+
+|===
+| Member Function | Description
+
+| bool ext_oneapi_can_access_peer(device peer,
+                                  ext::oneapi::peer_access value =
+                                  ext::oneapi::peer_access::access_supported)
+| Queries if this device may directly access the peer device's memory.
+Returns true if access is supported and `value` is defaulted to
+`ext::oneapi::peer_access::access_supported`. Returns true if `value` is
+`ext::oneapi::peer_access::access_enabled` and peer access has been enabled
+between this device and the peer device. Returns true if `value` is
+`ext::oneapi::peer_access::atomics_supported` and this device can perform atomic
+operations on the peer's memory. Returns false otherwise.
```

Comment:
I think this could be a little more clear about what the queries do.  In particular, we want it to be super clear whether a query returns true only when access is enabled with `peer` or if it returns true to indicate that access could potentially be enabled.  I like bulleted lists for things like this because they can easily be extended when we add new enum values.  Maybe something like:

> Queries the peer access status between this device and `peer` according to the query `value`:
>
> * `ext::oneapi::peer_access::access_supported`: Returns true only if it is possible for this device to enable peer access to USM device memory allocations located on the `peer` device.
>
> * `ext::oneapi::peer_access::access_enabled`: Returns true only if peer access is currently enabled from this device to the `peer` device.
>
> * `ext::oneapi::peer_access::atomics_supported`: When this query returns true, it indicates that this device may concurrently access and do atomic operations on USM device memory allocations located on the `peer` device when peer access is enabled to that device. [Some further text here describing the restriction when this query is false.]

Finally, what is the guarantee when `ext::oneapi::peer_access::atomics_supported` is *not* true?  Can the two devices still concurrently access the same USM device memory region (so long as they access disjoint elements), or must they  avoid concurrent access completely, as we say for USM host and shared memory?